File: DSA/Java/src/dsa/lib/math/ds/Matrix.java
package dsa.lib.math.ds;

public class Matrix {
    /**
     * LeetCode74
     * Binary Search. Treat the matrix like a 1d-array.
     * O(logM*N)
     *
     * @param matrix
     * @param target
     * @return
     */
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix.length == 0) return false;

        int m = matrix.length, n = matrix[0].length;
        int left = 0, right = m * n - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int pivot = matrix[mid / n][mid % n];

            if (pivot == target) return true;
            else if (pivot < target) left = mid + 1;
            else right = mid - 1;
        }
        return false;
    }
}

File: DSA/Go/dsa/lib/math/ds/matrix.go
package ds

/**
 * LeetCode74
 * Binary Search. Treat the matrix like a 1d-array.
 * O(logM*N)
 */
func searchMatrix(matrix [][]int, target int) bool {
	if len(matrix) == 0 {
		return false
	}

	m, n := len(matrix), len(matrix[0])
	left, right := 0, m*n-1
	for left <= right {
		mid := left + (right-left)/2
		pivot := matrix[mid/n][mid%n]

		if pivot == target {
			return true
		} else if pivot < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return false
}
